src/peer_connection.cpp

```diff
@@ -5460,23 +5460,22 @@ namespace libtorrent
 	int peer_connection::wanted_transfer(int channel)
 	{
 		TORRENT_ASSERT(is_single_thread());
-		shared_ptr<torrent> t = m_torrent.lock();
 
-		const int tick_interval = (std::max)(1, m_settings.get_int(settings_pack::tick_interval));
+#if TORRENT_NO_FPU
+		int const tick_interval = (std::max)(1, m_settings.get_int(settings_pack::tick_interval) / 1000);
+#else
+		float const tick_interval = m_settings.get_int(settings_pack::tick_interval) / 1000.f;
+#endif
 
 		if (channel == download_channel)
 		{
-			return std::max((std::max)(m_outstanding_bytes
-				, m_recv_buffer.packet_bytes_remaining()) + 30
-				, int(boost::int64_t(m_statistics.download_rate()) * 2
-					* tick_interval / 1000));
+			return std::max((std::max)(m_outstanding_bytes, m_recv_buffer.packet_bytes_remaining())
+				+ 30, static_cast<int>(m_statistics.download_rate() * tick_interval));
 		}
 		else
 		{
-			return std::max((std::max)(m_reading_bytes
-				, m_send_buffer.size())
-				, int((boost::int64_t(m_statistics.upload_rate()) * 2
-					* tick_interval) / 1000));
+			return std::max((std::max)(m_reading_bytes, m_send_buffer.size()),
+				static_cast<int>(m_statistics.upload_rate() * tick_interval));
```

Comment:
before you multiply by `tick_interval` you should widen the upload rate to a `boost::int64_t`. Also, the tick_interval is specified in milliseconds, you still need to divide by `1000` here

Reply:
I just noticed you divide it by 1000 before you assign `tick_interval`. This is problematic since the default interval is < 1 second, assigning it to an integer will give you 0.

Reply:
Nope, integer is evaluated as:
```int const tick_interval = (std::max)(1, m_settings.get_int(settings_pack::tick_interval) / 1000);```

The case you mentioned a `float` one.

Reply:
ok, so the tick_interval can never be less than 1 second. The existing code lets it be less than one second, which I think may be important

Reply:
Ah, I realized what's wrong with it. I will revert all this. But I still wish this function to be much simpler... Maybe the next PR will dedicated to this.